Support for the new Sailing trees

Scarcy opened the pull request:
I apologize if you don't want outside contributions to the plugin, but I've been loving it for a long time and wanted it to work for the new Sailing trees. Since the tree cutting times are now officially revealed, I added the trees with the despawn timers from the Wiki. 
The Rosewood Tree still only states that it's an estimated 274.8 Seconds, so I set it to be 275 seconds in the config. I cut down a few trees and the 275 second timer seemed to be accurate

<img width="272" height="225" alt="Camphor" src="https://github.com/user-attachments/assets/d5b054d2-0ba5-403a-a287-ff4b1cc68dc2" />
<img width="349" height="331" alt="RoseWood" src="https://github.com/user-attachments/assets/47410447-3fd0-4e71-99d6-b53c147f1769" />

Scarcy committed:
Added the timers for the three new Hardwoods that came with Sailing.

CreativeTechGuy commented:
This is great! Thank you so much for doing this. I didn't see the official announcement of the despawn timers for the new trees. Can you link the source you found for those numbers?

And can you also verify if there are any other ObjectIDs for these trees in-game? Eg: maybe some island has the same tree but it looks slightly different so would be a different model.

Scarcy commented:
> This is great! Thank you so much for doing this. I didn't see the official announcement of the despawn timers for the new trees. Can you link the source you found for those numbers?

Sure! The tree despawn timers are added to the specific wiki entry for each tree: [Camphor](https://oldschool.runescape.wiki/w/Camphor_tree), [Ironwood](https://oldschool.runescape.wiki/w/Ironwood_tree) and [Rosewood tree](https://oldschool.runescape.wiki/w/Rosewood_tree).

It seems now that the wiki has contradicting information on Ironwood and the Rosewood timers. The [Types of Trees](https://oldschool.runescape.wiki/w/Woodcutting#Types_of_trees) segment states a 390 tick timer for Ironwood and 440 tick timer for Rosewood. This is different to the timer on the individual wiki pages, where Ironwood claims a 400 tick timer and Rosewood a 458 tick timer.



> And can you also verify if there are any other ObjectIDs for these trees in-game? Eg: maybe some island has the same tree but it looks slightly different so would be a different model.

When I added the ObjectId for the trees, I looked at all the ObjectId's for the Camphor, Ironwood and Rosewood trees. I double checked again, and I can confirm that there's only one type for each of the new Sailing trees. For all three, there's one ObjectID for the normal Overworld trees and then many ObjectID's for the different Farming variation of the Hardwoods.

CreativeTechGuy commented:
The problem with the wiki reports is that they aren't 100% reliable since there's no surefire userland way to determine despawn timers. I was waiting until there was an official announcement from Jagex confirming the exact times. So we'd probably need to do a lot of tests to see if we can determine what the exact time is. Being off by 50+ ticks is pretty significant.

And just to confirm, when you said you looked at all ObjectIds, you mean you went around the world in-game to use the Identificator plugin on every tree with the same name? Because RuneLite may not have an enum for every unique instance of the tree. They often don't.

Scarcy commented:
> The problem with the wiki reports is that they aren't 100% reliable since there's no surefire userland way to determine despawn timers. I was waiting until there was an official announcement from Jagex confirming the exact times. So we'd probably need to do a lot of tests to see if we can determine what the exact time is. Being off by 50+ ticks is pretty significant.

Yeah that is very fair. I did some more testing on Camphor and Ironwood trees, and the Camphor timer looked to be correct. However, Ironwood was sometimes cut down when there was 5-6 seconds left on the timer (which seems to match with the 390 tick timer). 

I do understand wanting to wait for exact  despawn numbers :) Me personally as a big fan of the plugin would not mind it if it was +/- 5 seconds accurate, as the timer would at least give an indicator for how close the tree is to despawning. But not everybody will see it the same way

> And just to confirm, when you said you looked at all ObjectIds, you mean you went around the world in-game to use the Identificator plugin on every tree with the same name? Because RuneLite may not have an enum for every unique instance of the tree. They often don't.

I went by the Runelite API and with the Wiki page for ObjectID's. For Rosewoods and Ironwood, there's only one location that has them. The Camphor tree has multiple spots, but they are all on the same Island so I would assume that they have the same ObjectID. I can double check with the identification plugin when I get home :)

CreativeTechGuy commented:
Thank you for double checking all of that! If you can confirm the times within a few seconds and confirm there aren't any other random ObjectIDs, then I think this is good to go. Let me know when you are able to confirm. Appreciate your PR!

Scarcy commented:
> Thank you for double checking all of that! If you can confirm the times within a few seconds and confirm there aren't any other random ObjectIDs, then I think this is good to go. Let me know when you are able to confirm. Appreciate your PR!

I did some more detailed testing :)
Wrote a quick debug function to my local branch, that writes the despawn timers to a CSV file. I stood next to the Ironwoods and Rosewoods for 1-2 hours each, gathering stats on how long the trees lived.

Revised Numbers:
- Ironwood: 400 ticks -> 390 ticks (Sample size: 21 trees with > 25 players chopping)
- Rosewood: 458 ticks -> 440 ticks (Sample size: 10 trees with > 8 players chopping)

I also checked with the Identification plugin you recommended. I could not find any other variants than the standard tree variant for the three new trees.

Scarcy committed:
Revised Numbers after deeper testing

Did not get any valid testing data for Camphor, as there were no other players chopping those trees. The Camphor timer is currently just the states wiki timer.

Ironwood and Rosewood have been revised based on my research

Scarcy commented:
Also, thanks a lot for being open and supportive towards my contribution! I've wanted to dip my toes into plugin development, but it's always felt so daunting. So it's been very helpful to look at your clean code and implementations in this project

Scarcy committed:
Changed version from 1.4.8, to 1.5.0. Let me know if another version is more suitable to the change :)

CreativeTechGuy commented:
Wow great research. It's oddly suspicious that Ironwood has the exact same time as Magic and Rosewood as Redwood. That would make sense for those to be aligned from Jagex's side. But then the fact that Camphor is 6 seconds off of the next closest existing tree, Yew.

Unless your data strongly points otherwise, I'd be inclined to align Camphor with the Yew time since that seems to be the pattern. It'd be more strange for Camphor to be the odd one out and just by 6 seconds which is extremely small delta for a high level tree.

And then once we resolve that, I say we are good to merge!

Great contribution by the way. I look forward to seeing you participating in the RuneLite plugin community more now that you've gotten your feet wet!

Scarcy committed:
Changed Camphor to be the same as Yew timer

I definitely agree. It would make consistent sense that if Ironwood/Rosewood have the same/similar despawn rates to its closest tier tree, that Camphor has the same

Scarcy commented:
> Wow great research.
Thanks! It was pretty easy to extend your existing debug function `onGameObjectDespawned` to write data to a CSV file, and then use Python to get the stats 😄 

> Unless your data strongly points otherwise, I'd be inclined to align Camphor with the Yew time since that seems to be the pattern. It'd be more strange for Camphor to be the odd one out and just by 6 seconds which is extremely small delta for a high level tree.

I definitely agree. I did not trust any stats on the Camphor trees, unless I spend 10 hours cutting them. Ironwood trees had 100 people chopping on the forestry world, which makes the trees despawn closer to it's lifespawn. But I could not find worlds where more than 2 people chopped Camphor, so it would require a much larger sample size to get correct estimates.
 
> Great contribution by the way. I look forward to seeing you participating in the RuneLite plugin community more now that you've gotten your feet wet!

Thanks a lot! I gotta say, your code structure and implementation is extremely clean. I'm used to stepping into existing Java codebases that are so unorganized, not extendible and impossible to read without quizzing the code author lol. So massive props to you

CreativeTechGuy commented:
Thanks! That means a lot. I try to make it easy to understand and work with. :)

CreativeTechGuy commented:
You can track the release here: https://github.com/runelite/plugin-hub/pull/9796

Thank you so much for the contribution!
